CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,288 @@
+{
+    ComPtr<IToastNotificationFactory> factory;
+    auto hr = Windows::Foundation::GetActivationFactory(
+        HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotification).Get(),
+        &factory);
+    if (SUCCEEDED(hr))
+    {
+        hr = factory->CreateToastNotification(content, notification);
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::get_History(DesktopNotificationHistoryCompat *value)
+{
+    EnsureRegistered();
+
+    ComPtr<IToastNotificationManagerStatics> toastStatics;
+    HRESULT hr = Windows::Foundation::GetActivationFactory(
+        HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotificationManager).Get(),
+        &toastStatics);
+    if (SUCCEEDED(hr))
+    {
+        ComPtr<IToastNotificationManagerStatics2> toastStatics2;
+        hr = toastStatics.As(&toastStatics2);
+        if (SUCCEEDED(hr))
+        {
+            ComPtr<IToastNotificationHistory> history;
+            hr = toastStatics2->get_History(&history);
+            if (SUCCEEDED(hr))
+            {
+                *value = DesktopNotificationHistoryCompat(s_aumid, history);
+            }
+        }
+    }
+
+    return S_OK;
+}
+
+bool DesktopNotificationManagerCompat::CanUseHttpImages()
+{
+    return IsRunningAsUwp();
+}
+
+void DesktopNotificationManagerCompat::EnsureRegistered()
+{
+    // If not registered AUMID yet
+    if (!s_registeredAumidAndComServer)
+    {
+        // Check if Desktop Bridge
+        if (IsRunningAsUwp())
+        {
+            // Implicitly registered, all good!
+            s_registeredAumidAndComServer = true;
+        }
+        else
+        {
+            // Otherwise, incorrect usage
+            throw std::exception("You must call RegisterAumidAndComServer first.");
+        }
+    }
+
+    // If not registered activator yet
+    if (!s_registeredActivator)
+    {
+        // Incorrect usage
+        throw std::exception("You must call RegisterActivator first.");
+    }
+}
+
+bool DesktopNotificationManagerCompat::IsRunningAsUwp()
+{
+    if (!s_hasCheckedIsRunningAsUwp)
+    {
+        s_hasCheckedIsRunningAsUwp = true;
+
+        // https://stackoverflow.com/questions/39609643/determine-if-c-application-is-running-as-a-uwp-app-in-desktop-bridge-project
+        UINT32 length;
+        wchar_t packageFamilyName[PACKAGE_FAMILY_NAME_MAX_LENGTH + 1];
+        LONG result = GetPackageFamilyName(GetCurrentProcess(), &length, packageFamilyName);
+        s_isRunningAsUwp = result == ERROR_SUCCESS;
+    }
+
+    return s_isRunningAsUwp;
+}
+
+DesktopNotificationHistoryCompat::DesktopNotificationHistoryCompat(const wchar_t *aumid, ComPtr<IToastNotificationHistory> history)
+{
+    m_aumid = aumid;
+    m_history = history;
+}
+
+DesktopNotificationHistoryCompat::DesktopNotificationHistoryCompat()
+{
+    // TODO: This constructor should be removed, if we made this all support ComPtr this would probably not be necessary?
```

Comment:
I'm not sure what the ComPtr comment is about, but if the default constructor is not needed then delete it.

Reply:
Ideally the History object should be able to be used as a ComPtr<History>, right?

However, in the consuming app code, if I placed it as a ComPtr, there was a compile error about the Release method not being implemented, since I think ComPtr capable classes need to extend some RuntimeClass? And then that seems complicated...

So in order to initialize it from app code, I had to first declare the object, which requires the default constructor, and then grab it. Maybe as part of moving to throwing exceptions, this could be inlined, and the default constructor wouldn't be needed?

```cpp
    DesktopNotificationHistoryCompat history;
    auto hr = DesktopNotificationManagerCompat::get_History(&history);
    if (SUCCEEDED(hr))
    {
        // Clear all toasts
        hr = history.Clear();
    }
```

Reply:
Unless you have a reason, you should not make this a ComPtr. std::shared_ptr works just as well and it's less arcane. 

You are correct, If DNM::get_History returned a DNHC directly you would not need a default constructed object.